### ISSUE TITLE
Add 'description' to fields and models

### ISSUE THREAD

**Issue by chadrik** [schematics/schematics#323]
Descriptions are a fundamental aspect of a well-designed schema.  They are not only great documentation for other developers, they can be programmatically inspected to populate UIs, etc. They're a key part of the JSON Schema specification and so their addition to `schematics` would enable tools like [jsonschematics](https://github.com/voidfiles/jsonschematics) to create more compliant output, which would in turn make it more compatible with other tools that take JSON Schema documents as an input.

For types, the description makes sense as a new argument to `BaseType.__init__`.  For `Model`s, the value could be taken from the object's `__doc__` attribute (or perhaps just the first sentence).

Thanks for listening.

**Comment by kracekumar** [comment]
It makes sense to me. At some point I want to generate schema out of schematics to share with `mobile` developers. The case I am speaking is `Flask + Schematics (validation for user input)` for developing API's.

+1 from my side for the idea.

**Comment by bintoro** [comment]
As far as types go, isn't this a duplicate of #333?

@smcclstocks If you'd like to submit another PR that incorporates the changes discussed in #333, I'm certain we can get it merged. And in case you want to implement the dotted syntax (`annotations.foobar`), I've added the `DataObject` class to `datastructures.py` that can be used for this.

**Comment by bintoro** [comment]
Would it make sense to put model annotations in the `Options` class or in a designated class variable? This seems more flexible than the docstring approach.

**Comment by chadrik** [comment]
I've thought about this a bit and I don't think that the description should be part of annotations (#333). Annotations should be for *arbitrary* and *domain-specific* metadata, like the `as_sql_col` example, whereas, the description of a field is fairly universal.  If we choose to relegate this to "annotations", then we miss the opportunity to define a standard:  some users will choose "description", while others choose "help" or "doc".  This will undermine the stated goal of allowing schematics to serve as a universal model for converting to other standards, like jsonschema.

**Comment by bintoro** [comment]
@chadrik Generating a schema, or any other usage of metadata such as `description`, is still going to require some kind of specification that describes the available properties. Since nothing of the sort exists yet, we're free to set any convention we like. This includes suggesting specific semantics for certain items within `annotations`.

If we were to separate domain-specific annotations from more universal metadata, is `description` all we'll ever need? Before storing such properties directly on the field instance, we should be reasonably sure that it won't encroach on the `__dict__` namespace too much. For example, an attribute called `"type"` could be a realistic metadata descriptor. But I would be uncomfortable reserving `BaseType.type` for that kind of purpose.

So, it's possible we should actually store the metadata properties in a container, which leads to the question of whether it's overkill to have *separate* containers for annotations and standard metadata, if the latter will *typically* only contain one or two items?

Also, on models, a `description` attribute directly on the class is not viable anyway, so it'd have to go in `_options` at minimum.

**Comment by bintoro** [comment]
Looking at JSON Schema, it seems we can distinguish a few kinds of properties on fields:

* Instance-specific properties that are purely metadata:
  * title
  * description

* Instance-specific properties that have functional meaning:
  * `maximum` for `NumberType`
  * `maxLength` for `StringType`

* Type-specific properties:
  * `type` (as in the primitive type of the Schematics type)
  * `maxLength` for `SHA1Type`

Anything that has functional meaning already exists as an attribute, either on the class or the instance.

`type` would be set on the class in some standardized way:

```python
class UUIDType(BaseType):
    __native__ = uuid.UUID
    __primitive__ = str
```

So, as far as JSON Schema is concerned, the only question is about the pure-metadata properties `title` and `description`. If it's just these two, what's the drawback of putting them in `annotations` as opposed to a dedicated dictionary, say, `meta`?

Any other standardized specifications we should take into account?

**Comment by chadrik** [comment]
It would be confusing to have two metadata dictionaries, `meta` and `annotations`, and I think we are in agreement about that, so the main question is whether universal metadata should go in annotations or on instance attributes.  They would be the first attributes to gain this distinction that do not play a functional role within schematics itself, so you rightly want to be careful about setting a precedent that could lead down a slippery slope.  I believe that `title` and `description` covers all of our needs, but I haven't done an exhaustive search through other schema-schemas.

> So, as far as JSON Schema is concerned, the only question is about the pure-metadata properties title and description. If it's just these two, what's the drawback of putting them in annotations as opposed to a dedicated dictionary, say, meta?

The main drawback is that we lose the self-documenting nature of a keyword argument.  If we want to encourage users of schematics to use `title` and `description` over other alternatives, a keyword argument is a great way to do so. 

Ultimately, the biggest practical difference will be the clarity of the documentation and the schema definitions.

## Documenation

### As Attributes

```
    :param annotations:
        Dictionary for storing custom metadata associated with the object.  These have
        no functional role within schematics.
    :param title:
        Brief title of the field or model.  Used by external tools for human-readable labels
    :param description:
        Description of the field or model.  Used by external tools for help, tooltips, etc.
```
And of course, `BaseType.__init__` will include `title` and `description` keywords.

### As Annotations
```
    :param annotations:
        Dictionary for storing custom metadata associated with the object.  These have
        no functional role within schematics.
        To encourage compatibility with external tools, we suggest these keys 
        for common annotations:
        title:
            Brief title of the field or model.  Used for human-readable labels.
        description:
            Description of the field or model.  Used for help, tooltips, etc.
```

## Schema Definition

### As Attributes / Keywords

```python
class Book(Model):
    class Options:
        title = 'A Book'
        description = 'A book stored at your local library'

    author = StringType(
        required=True, 
        description='The person that wrote the thing')
    ddid = StringType(
         required=True,
         title='Dewey Decimal ID',
         description='ID using the Dewey Decimal Classification System')
```

### As Annotations
```python
class Book(Model):
    class Options:
        annotations = {
            'title': 'A Book',
            'description': 'A book stored at your local library'
        }

    author = StringType(
        required=True, 
        annotations={
            'description': 'The person that wrote the thing'
        })
    ddid = StringType(
         required=True,
         annotations={
            'title': 'Dewey Decimal ID',
            'description': 'ID using the Dewey Decimal Classification System'
        })
```

In making this, I'm not trying to sell you on one or the other.  I'm just trying to comprehend the practical outcome of this decision.  As someone who will be using this feature, I prefer the look of model definitions using direct keywords instead of annotations dictionaries, but ultimately, I'll be happy with either one.

**Comment by smcclstocks** [comment]
Attributes are easier for me as well but it would be great if we could "break" out when needed with something like a dict for arbitrary keywords. For instance, maybe I want to have a property on certain fields that would disallow modification once initially set called: 'allow_update=False'. So although title & description seem like they should be first class citizens & I agree they should, I don't want to overlook the value of a "safe" way to have custom keywords without having to sub-class every type.

I could think of at least a dozen different attributes that would be nice to define on a models field but they could be 100% specific to my API's & I don't want to bloat schematics with all that. So I guess what I'm saying is I'd like to have both :)   ... annotations or some dict of arbitrary key/vals & a top-level title/description.

**Comment by bintoro** [comment]
@smcclstocks We're definitely going to have a container where you can put anything. The only question now is whether to reserve the `title` and `description` slots from said container or put them on the field as attributes in their own right.

A type wanting to use an attribute called `title` or `description` seems like a remote prospect, so I'm not concerned about that. It's just that these two deinitely feel like annotations to me.

But if everybody else agrees that first-class attributes are the way to go, I'm not opposed.

**Comment by chadrik** [comment]
Honestly, I could go either way on this.  Since #333 is the superset of the two issues, I think it would be good to start with that feature and give this one more time to stew.  I'm now going to throw a few suggestions into the other issue.

**Comment by sheilatron** [comment]
When this feature is ready, maybe it could provide the basis for a Sphinx extension. It would be nice to be able provide a docstring directive that would work with autodoc. As it is, I'm currently just putting all the descriptions into the docstring of the model class.

I've never worked on a Sphinx extension but might give it a try if the idea proves worthwhile.

**Comment by chadrik** [comment]
PR #333 was recently merged, so there may no longer be the need to implement this feature.  

@sheilatron you can now place your docstrings in a 'description' metadata key for each field.  If you build a sphinx extension it would go a long way to establishing that key as a convention for others to use.

**Comment by bintoro** [comment]
A purpose-built Sphinx extension would be pretty dope!

As for this issue, is there anything left to do here, or was this fully resolved by #390?

**Comment by chadrik** [comment]
#390 is good for me.
